### PR TITLE
Switch to Google Proto

### PR DESCRIPTION
**Pull request by neekolas**
## Summary
I was tasked by @jazzz to help out with some Protobuf encoding issues. This solution is not my favorite, but I'm struggling to come up with a less bad option.

Here's the background:
- `go-waku` uses `gogoproto` instead of the Google Protobuf library to generate code
- `gogoproto` is [currently unmaintained](https://github.com/gogo/protobuf/issues/691) and doesn't support newer versions of Protobuf, which our protos implicitly require. When trying to compile our protos using `gogoproto` we get an error. Even if we could rewrite our protos to avoid the error, it still seems unwise to hitch our wagon to a library that is completely dead.
- Waku also uses a library called `protoio` to read and write protobufs to length delimited lib2p2 streams. That library is tied to `gogoproto` types. 

This is a fork of `protoio` that is compatible with vanilla Google Protobuf

## Alternatives Considered
- Use https://github.com/CrowdStrike/csproto and figure out how to make it work with `protoio`
  - `csproto` is not a complete replacement for Google Protobuf and we would still have to rely on the Google Protobuf message interface
  - Main advantages are performance, which aren't our primary concern here. Much less supported and less active Github.
  - Still could drop this in later with minimal impact
  - Still requires the same work to implement with `protoio`
- Completely rewrite `protoio` in `xmtp-node-go`
  - Fair bit of work and would likely just recreate what LibP2P has already done here
- Patch `gogoproto` to fix our immediate problem
   - The issues arise from a major breaking version change in the Google library, which would be a significant task to keep up with. The team literally chose to abandon the project rather than try and keep up with Google
- Use an old version of `gogoproto` which apparently works.
   - Feels like we are going to wind up needing an important security update or bug fix at some point and will be trapped. 

## Notes
- I had to comment out some code that relies on `gogoproto`'s `MarshalTo` method. Their code already provides a fallback for types that do not have a `MarshalTo` method but I have no idea what the implications are for the fallback
- The original test suite works, but with some modifications to be compatible with Google Protobuf library

### CODE REVIEW

File: protoio/uvarint_test.go [hunk 1]
```diff
@@ -148,3 +141,189 @@ func iotest(writer protoio.WriteCloser, reader protoio.ReadCloser) error {
 	}
 	return nil
 }
+
+// Everything below this line is borrowed from the generated test code in gogoproto
```
Comment (neekolas) [inline protoio/uvarint_test.go:145]: `gogoproto` has some magic options to generate test helpers. Now that we are not using `gogoproto` I had to recreate their test fixtures, generate the code with Google Protobuf, and then copy and paste over the test helpers below. It's not pretty.

File: test/test.proto [hunk 2]
```diff
@@ -0,0 +1,22 @@
+syntax = "proto3";
+
+package test;
+option go_package = "github.com/libp2p/go-msgio/test";
+
+message NinOptNative {
```
Comment (neekolas) [inline test/test.proto:6]: This message is borrowed from `gogoproto`'s test suite

File: protoio/uvarint_writer.go [hunk 3]
```diff
@@ -62,25 +62,26 @@ func (uw *uvarintWriter) WriteMsg(msg proto.Message) (err error) {
...
+	// 		}
+	// 		lenOff := varint.PutUvarint(uw.buffer, uint64(n))
+	// 		_, err = m.MarshalTo(uw.buffer[lenOff:])
+	// 		if err != nil {
+	// 			return err
+	// 		}
+	// 		_, err = uw.w.Write(uw.buffer[:lenOff+n])
+	// 		return err
+	// 	}
+	// }
```
Comment (jazzz) [inline protoio/uvarint_writer.go]: After parity has been shown, Is there a reason to keep the block around?

Comment (neekolas) [reply protoio/uvarint_writer.go]: Not really. I can delete.

**Review by jazzz — APPROVED** [review]
:+1:

**Comment by mkobetic** [comment]
> I had to comment out some code that relies on `gogoproto`'s `MarshalTo` method. Their code already provides a fallback for types that do not have a `MarshalTo` method but I have no idea what the implications are for the fallback

It's not clear to me when is the uvarint reader/writer used, but assuming it's in all general length-prefixed cases, then the blast radius can be quite wide. It seems that `MarshalTo` might be intended as a more optimized version of `proto.Marshal`, but looking at the code it still just seems to write into an internal buffer and then copy the result into the writer, so I'm not sure what would be the optimization here. In general case it could be possible that `MarshalTo` produces different encoding than `proto.Marshal`, in that case the impact could be quite bad, as things that implement MarshalTo would start encoding differently. Do you intend to submit this upstream? I suspect there might be pushback.

**Comment by neekolas** [comment]
@mkobetic no intention to push this upstream. Changing the underlying proto library would be a huge breaking change for all users of the library.

**Comment by neekolas** [comment]
FWIW, their existing test suite all passes with my changes, so I'm fairly confident that the methods are functionally equivalent for encoding/decoding.

**Review by mkobetic — APPROVED** [review]

**Pull request merged**